ext/date/php_date.c

```diff
@@ -2380,6 +2383,34 @@ static zend_object *date_object_clone_timezone(zval *this_ptr) /* {{{ */
 	return &new_obj->std;
 } /* }}} */
 
+static int date_object_compare_timezone(zval *tz1, zval *tz2) /* {{{ */
+{
+	php_timezone_obj *o1, *o2;
+
+	o1 = Z_PHPTIMEZONE_P(tz1);
+	o2 = Z_PHPTIMEZONE_P(tz2);
+
+	if (!o1->initialized || !o2->initialized) {
+		php_error_docref(NULL, E_WARNING, "Trying to compare an incomplete DateTimeZone object");
+		return 1;
+	}
+
+	if (o1->type != o2->type) {
+		php_error_docref(NULL, E_WARNING, "Trying to compare different kinds of DateTimeZone objects");
+		return 1;
```

Comment:
Technically you can compare a `TIMELIB_ZONETYPE_OFFSET` and `TIMELIB_ZONETYPE_ABBR` kind, as they will have both the `utc_offset` (and `dst` flag) set, but it's probably best not to do.